src/plugins/intel_gpu/src/graph/network.cpp

```diff
@@ -876,8 +877,30 @@ void network::allocate_primitive_instance(program_node const& node) {
         if (node.is_type<data>())
             _data_outputs.push_back(inst);
     }
-    if (node.is_constant())
-        transfer_memory_to_device(inst, node);
+
+    auto transfer_memory_to_host = [&](std::shared_ptr<primitive_inst> instance, program_node const& node) {
+        auto& inst_mem = instance->output_memory();
+        auto alloc_type = inst_mem.get_allocation_type();
+
+        if (alloc_type == allocation_type::usm_device) {
+            // Allocate and transfer memory
+            auto host_mem = inst_mem.get_engine()->allocate_memory(inst_mem.get_layout(), allocation_type::usm_host, false);
+            host_mem->copy_from(get_stream(), inst_mem);
+            _memory_pool->release_memory(&inst_mem, node.id(), get_id());
+            instance->set_output_memory(host_mem);
+        }
+    };
+
+    if (node.is_constant()) {
+        auto users = node.get_users();
+        if (users.size() == 1
+            && users.front()->is_type<reshape>()
+            && users.front()->is_dynamic()) {
+                transfer_memory_to_host(inst, node);
+        } else {
+            transfer_memory_to_device(inst, node);
+        }
+    }
```

Comment:
Instead of doing this here, how about do this at "transfer_memory" stage of program build?

Reply:
Fixed it